### PR TITLE
Rewrite `min_max_string` and `min_max_binary`

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>

# Changes in the PR
1. Replace `fold` with `reduce`
2. use macro to abstract the common expression
3. Add test to cover the `no null` cases.
4. Add benchmark for `min utf8`
5. Less code is better
6. Functional style

# No obvious performance impact
```
min 2^10 utf8           time:   [33.766 ns 33.786 ns 33.810 ns]                           
                        change: [+0.0167% +0.2385% +0.4298%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

min null 2^10 utf8      time:   [33.352 ns 33.370 ns 33.391 ns]                                
                        change: [-4.7561% -4.6289% -4.5317%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe
```

### CODE REVIEW

File: benches/aggregate.rs [hunk 1]
```diff
@@ -42,6 +42,18 @@ fn add_benchmark(c: &mut Criterion) {
         c.bench_function(&format!("min null 2^{} f32", log2_size), |b| {
             b.iter(|| bench_min(&arr_a))
         });
+
+        let arr_a = create_string_array::<i32>(1, size, 0.0, 0);
```
Comment (HaoYang670) [inline benches/aggregate.rs:46]: I set the length of each string to be 1 to make the time of comparing each element to be determined.

**Comment by jorgecarleitao** [comment]
Thanks a lot for this PR.

I went through this and also ran the benches on main vs this PR with different string lengths and array lengths and the performance is equivalent on my machine.

Given that this is less code and more tested, we ship it :)
